simde/x86/avx512f.h

```diff
@@ -3780,6 +3780,92 @@ simde_mm512_maskz_sub_pd(simde__mmask8 k, simde__m512d a, simde__m512d b) {
   #define _mm512_maskz_sub_pd(k, a, b) simde_mm512_maskz_sub_pd(k, a, b)
 #endif
 
+SIMDE__FUNCTION_ATTRIBUTES
+simde__m512i
+simde_mm512_slli_epi32 (simde__m512i a, unsigned int imm8) {
+  simde__m512i_private
+    r_,
+    a_ = simde__m512i_to_private(a);
+
+  /* The Intel Intrinsics Guide says that only the 8 LSBits of imm8 are
+   * used.  In this case we should do "imm8 &= 0xff".  However in
+   * practice all bits are used. */
+
+  /* The 128-bit and 256-bit versions don't seem to test for this and run-tests fails.
+   * I think that this is an error and that they should perform this test. */
+  if (imm8 > 31) {
+    simde_memset(&r_, 0, sizeof(r_));
+  } else {
+    #if defined(SIMDE_ARCH_X86_AVX2)
+      r_.m256i[0] = simde_mm256_slli_epi32(a_.m256i[0], HEDLEY_STATIC_CAST(int, imm8));
+      r_.m256i[1] = simde_mm256_slli_epi32(a_.m256i[1], HEDLEY_STATIC_CAST(int, imm8));
+    #elif defined(SIMDE_ARCH_X86_SSE2)
+      r_.m128i[0] = simde_mm_slli_epi32(a_.m128i[0], HEDLEY_STATIC_CAST(int, imm8));
+      r_.m128i[1] = simde_mm_slli_epi32(a_.m128i[1], HEDLEY_STATIC_CAST(int, imm8));
+      r_.m128i[2] = simde_mm_slli_epi32(a_.m128i[2], HEDLEY_STATIC_CAST(int, imm8));
+      r_.m128i[3] = simde_mm_slli_epi32(a_.m128i[3], HEDLEY_STATIC_CAST(int, imm8));
+    #elif defined(SIMDE_VECTOR_SUBSCRIPT_SCALAR)
+      r_.u32 = a_.u32 << imm8;
+    #else
+      SIMDE__VECTORIZE
+      for (size_t i = 0 ; i < (sizeof(r_.u32) / sizeof(r_.u32[0])) ; i++) {
+        r_.u32[i] = a_.u32[i] << imm8;
+      }
+    #endif
+  }
+
+  return simde__m512i_from_private(r_);
+}
+#if defined(SIMDE_AVX512F_NATIVE)
+  #define simde_mm512_slli_epi32(a, imm8) _mm512_slli_epi32(a, imm8)
+#endif
+#if defined(SIMDE_AVX512F_ENABLE_NATIVE_ALIASES)
+  #define _mm512_slli_epi32(a, imm8) simde_mm512_slli_epi32(a, imm8)
+#endif
+
+SIMDE__FUNCTION_ATTRIBUTES
+simde__m512i
+simde_mm512_slli_epi64 (simde__m512i a, unsigned int imm8) {
+  simde__m512i_private
+    r_,
+    a_ = simde__m512i_to_private(a);
+
+  /* The Intel Intrinsics Guide says that only the 8 LSBits of imm8 are
+   * used.  In this case we should do "imm8 &= 0xff".  However in
+   * practice all bits are used. */
+
+  /* The 128-bit and 256-bit versions don't seem to test for this and run-tests fails.
+   * I think that this is an error and that they should perform this test. */
```

Comment:
I haven't tested this specific function, but in general the problem seems to be a mismatch in expectations.

In assembly, values outside the expected range ([0, 63] in this case) are either a compile-time error or are masked off by the CPU, I'm not entirely sure which.

The IIG specifies that the high bits are masked off, but that seems to just be a shorthand to say that the values must fall in the expected range.

Some compilers generate errors if the values don't fall in the expected range; clang tends to be more aggressive about this, but I've seen GCC do it, too.

Some compilers also allow values outside the expected range as sort of an extension.  The problem here is that there is no consistency across compilers.  I've run into situations where MSVC generates a different result from GCC/clang for these.

My position in SIMDe has basically been that values outside the expected range are undefined behavior, and we should add `HEDLEY_REQUIRE_MSG` macros to the functions to try to generate a compile time warning.

Reply:
> Some compilers generate errors if the values don't fall in the expected range; clang tends to be more aggressive about this, but I've seen GCC do it, too.
> 
> Some compilers also allow values outside the expected range as sort of an extension. The problem here is that there is no consistency across compilers. I've run into situations where MSVC generates a different result from GCC/clang for these.
> 

While compilers may test in the case of compile-time constants I doubt that they would insert a run-time test. This would slow down code just to handle a case that probably shouldn't happen (if it does it probably indicates a bug in the code).

> My position in SIMDe has basically been that values outside the expected range are undefined behavior, and we should add `HEDLEY_REQUIRE_MSG` macros to the functions to try to generate a compile time warning.
> 

Yes, I agree. Not only are there differences between compilers but it is also possible that Intel and AMD processors treat these grey areas differently (or even different generations of Intel processors).
However in our previous discussion around here https://github.com/nemequ/simde/pull/150#discussion_r403509435 you suggested that I introduce tests for out of range values.
If we now agree to treat these as undefined, should I remove these tests?
In fact if you add the Hedley macro to generate a warning, this will cause failure on compilations which treat warnings as errors and so I MUST remove the tests.
However there is one out of range value, the exact size of the type, which I feel is useful (see below). Should I at least test this?

> _Edit_: it's also worth noting that right shifting a signed integer by more than the width of the type is undefined behavior in C.
> 

IIRC in K&R it was even worse than this:
- ANY right shift of an signed type was undefined. I guess that this may have been because the DEC PDP11 (for which they developed C IIRC) may only have had instructions for logical right shifts. However luckily all machines I have used generate an arithmetic right shift.
- Any shift by the exact width of the type was also undefined. Indeed many machines take the shift modulo the width of the type. This means that a UINT32 shifted by 32 bits doesn't disappear to zero but is unchanged.

I remember being once bitten by this. I wrote a left rotate of a UINT32 u by n bits as:
r = (u << n) | (u >> (32 - n));
This fails for n = 0 as it generates a 32-bit right shift which generally leaves u unchanged instead disappearing to zero.

I hope that this has been improved since K&R because arithmetic right shifts are very useful and a shift by the exact word size can be useful too.

Reply:
> However in our previous discussion around here #150 (comment) you suggested that I introduce tests for out of range values.

You're right, that does seem inconsistent.  The issues you encountered helped convince me to treat these as undefined, which is why I suggested what I did here.  However, as I mentioned in [another comment](https://github.com/nemequ/simde/pull/253#issuecomment-621489223), after looking into this a bit more I think that for this particular function it actually makes sense to support values outside the expected range.

Take a look at <https://godbolt.org/z/_5Cku9>.

 * If `imm8` is within the expected ([0, 63]) range it's what you'd expect the instruction for this intrinsic to be: `vpslld zmm, k, zmm, i8`.  
 * If `imm8` is in [64, 127], MSVC will continue to generate the same code as the version in the expected range.  That's fine; on x86/x86_64 we know the result will always be 0.
 * If `imm8` is > 127, MSVC will load the int into the first lane of an `__m128i` and do a `_mm512_sll_epi32`.  For values > 64 the result will always be zero.
 * Knowing that the result will alwasy be 0, Clang, GCC, and ICC will all just xor the input with itself (the same thing `_mm512_setzero_si512` does) if `imm8` > 63.

So basically, *nobody*, not even Intel, actually does what the Intrinsics Guide says.  But everyone does generate the same results.

I think we're just going to have to treat these on a case-by-case basis where if compilers actually agree we can operate that way in SIMDe.

> While compilers may test in the case of compile-time constants I doubt that they would insert a run-time test. This would slow down code just to handle a case that probably shouldn't happen (if it does it probably indicates a bug in the code).

They don't really test.  That Compiler Explorer link above also has a "unknown" version if you want to see for yourself.  If they don't know the value at compile time the compiler will just use `_mm512_sll_epi16(a, _mm_cvtsi32_si128(imm8))` instead.

The problem is that the immediate version is *way* faster.  On Icy Lake, just for the shift we're talking about a latency of up to 8 cycles vs 1 for the immediate mode.

> IIRC in K&R it was even worse than this:

Ouch!  I didn't realize it was that bad :(